File: src/main/java/AoC2022_16.java
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toSet;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.graph.AStar;
import com.github.pareronia.aoc.graph.AStar.Result;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

import lombok.RequiredArgsConstructor;

public class AoC2022_16 extends AoCBase {
    
    private final String[] valves;
    private final int[] rates;
    private final Set<Integer>[] tunnels;
    private int start;
    
    @SuppressWarnings("unchecked")
    private AoC2022_16(final List<String> input, final boolean debug) {
        super(debug);
        final int size = input.size();
        this.valves = new String[size];
        this.rates = new int[size];
        this.tunnels = new Set[size];
        final Map<String, Set<String>> edges = new HashMap<>();
        final Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < size; i++) {
            final String line = input.get(i).replaceAll("[,;]", "");
            final String[] splits = line.split(" ");
            final String name = splits[1];
            if ("AA".equals(name)) {
                this.start = i;
            }
            valves[i] = name;
            map.put(name, i);
            final int rate = Integer.parseInt(splits[4].split("=")[1]);
            rates[i] = rate;
            IntStream.range(9, splits.length).mapToObj(j -> splits[j])
                .forEach(v -> edges.computeIfAbsent(name, k -> new HashSet<>()).add(v));
        }
        edges.entrySet().forEach(e -> {
            final int idx = map.get(e.getKey());
            this.tunnels[idx] = e.getValue().stream().map(map::get).collect(toSet());
        });
        trace(() -> "valves: " + IntStream.range(0, this.valves.length)
                .mapToObj(i -> String.format("%d: %s", i, this.valves[i]))
                .collect(joining(", ")));
        trace(() -> "rates: " + IntStream.range(0, this.rates.length)
                .mapToObj(i -> String.format("%d: %s", i, this.rates[i]))
                .collect(joining(", ")));
        trace(() -> "tunnels: " + IntStream.range(0, this.tunnels.length)
                .mapToObj(i -> String.format("%d: %s", i, this.tunnels[i]))
                .collect(joining(", ")));
    }
    
    public static final AoC2022_16 create(final List<String> input) {
        return new AoC2022_16(input, false);
    }

    public static final AoC2022_16 createDebug(final List<String> input) {
        return new AoC2022_16(input, true);
    }
    
    @RequiredArgsConstructor
    private final class DFS {
        private final int[][] distances;
        private final int maxTime;
        private final boolean sample;
        private final Map<Long, Integer> bestPerUsed = new HashMap<>();
        private long used = 0L;
        private int maxFlow = 0;
        private int cnt = 0;
        
        public void dfs(final int start, final int time) {
            cnt++;
            for (int i = 0; i < valves.length; i++) {
                final long idx = 1L << i;
                if (rates[i] == 0 || (used & idx) != 0) {
                    continue;
                }
                final int newTime = time + 1 + distances[start][i];
                if (newTime < maxTime) {
                    final int flow = rates[i] * (maxTime - newTime);
                    if (!this.sample && maxFlow + flow
                            < bestPerUsed.getOrDefault(used + idx, 0)) {
                        continue;
                    }
                    maxFlow += flow;
                    used += idx;
                    dfs(i, newTime);
                    used -= idx;
                    maxFlow -= flow;
                }
            }
            bestPerUsed.merge(used, maxFlow, Math::max);
        }
    }

    private int[][] getDistances() {
        final int[] relevantValves = IntStream.range(0, this.valves.length)
                .filter(i -> this.rates[i] > 0 || i == this.start)
                .toArray();
        final int size = valves.length;
        final int[][] distances = new int[size][size];
        for (final int i : relevantValves) {
            final Result<Integer> result = AStar.execute(
                    i,
                    v -> false,
                    v -> this.tunnels[v].stream(),
                    v -> 1);
            for (final int j : relevantValves) {
                distances[i][j] = (int) result.getDistance(j);
            }
        }
        trace(() -> "    " + Arrays.stream(relevantValves)
                .mapToObj(i -> this.valves[i]).collect(joining(" | ")));
        Arrays.stream(relevantValves).forEach(i -> {
            trace(() -> this.valves[i] + "  "
                    + Arrays.stream(relevantValves)
                            .mapToObj(j -> String.format("%2d", distances[i][j]))
                            .collect(joining(" | ")));
        });
        return distances;
    }

    @Override
    public Integer solvePart1() {
        final int[][] distances = getDistances();
        final DFS dfs = new DFS(distances, 30, false);
        dfs.dfs(this.start, 0);
        log("dfs: " + dfs.cnt);
        return dfs.bestPerUsed.values().stream()
                .mapToInt(Integer::valueOf)
                .max().orElseThrow();
    }
    
    private int solve2(final boolean sample) {
        final int[][] distances = getDistances();
        final DFS dfs = new DFS(distances, 26, sample);
        dfs.dfs(this.start, 0);
        log("dfs: " + dfs.cnt);
        return dfs.bestPerUsed.entrySet().stream()
            .flatMapToInt(e1 -> dfs.bestPerUsed.entrySet().stream()
                .filter(e2 -> (e1.getKey() & e2.getKey()) == 0)
                .mapToInt(e2 -> e1.getValue() + e2.getValue()))
            .max().orElseThrow();
    }

    @Override
    public Integer solvePart2() {
        return solve2(false);
    }
    
    public static void main(final String[] args) throws Exception {
        assert AoC2022_16.createDebug(TEST).solvePart1() == 1651;
        assert AoC2022_16.createDebug(TEST).solve2(true) == 1707;

        final Puzzle puzzle = Aocd.puzzle(2022, 16);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_16.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_16.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST = splitLines(
        "Valve AA has flow rate=0; tunnels lead to valves DD, II, BB\r\n" +
        "Valve BB has flow rate=13; tunnels lead to valves CC, AA\r\n" +
        "Valve CC has flow rate=2; tunnels lead to valves DD, BB\r\n" +
        "Valve DD has flow rate=20; tunnels lead to valves CC, AA, EE\r\n" +
        "Valve EE has flow rate=3; tunnels lead to valves FF, DD\r\n" +
        "Valve FF has flow rate=0; tunnels lead to valves EE, GG\r\n" +
        "Valve GG has flow rate=0; tunnels lead to valves FF, HH\r\n" +
        "Valve HH has flow rate=22; tunnel leads to valve GG\r\n" +
        "Valve II has flow rate=0; tunnels lead to valves AA, JJ\r\n" +
        "Valve JJ has flow rate=21; tunnel leads to valve II"
    );
}

File: src/main/python/AoC2022_16.py
#! /usr/bin/env python3
#
# Advent of Code 2022 Day 16
#


from __future__ import annotations

import sys
from collections import defaultdict
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.graph import Dijkstra

TEST = """\
Valve AA has flow rate=0; tunnels lead to valves DD, II, BB
Valve BB has flow rate=13; tunnels lead to valves CC, AA
Valve CC has flow rate=2; tunnels lead to valves DD, BB
Valve DD has flow rate=20; tunnels lead to valves CC, AA, EE
Valve EE has flow rate=3; tunnels lead to valves FF, DD
Valve FF has flow rate=0; tunnels lead to valves EE, GG
Valve GG has flow rate=0; tunnels lead to valves FF, HH
Valve HH has flow rate=22; tunnel leads to valve GG
Valve II has flow rate=0; tunnels lead to valves AA, JJ
Valve JJ has flow rate=21; tunnel leads to valve II
"""


class Cave(NamedTuple):
    start: int
    valves: list[str]
    rates: list[int]
    tunnels: list[set[int]]

    @classmethod
    def from_input(cls, lines: list[str]) -> Cave:
        valves = [""] * len(lines)
        the_map = dict[str, int]()
        rates = [0] * len(lines)
        edges = defaultdict[str, set[str]](set)
        tunnels = [set[int]()] * len(lines)

        for i, line_ in enumerate(lines):
            line = line_.replace(",", "")
            line = line.replace(";", "")
            _, name, _, _, rate, _, _, _, _, *rest = line.split()
            if name == "AA":
                start = i
            valves[i] = name
            the_map[name] = i
            rates[i] = int(rate.split("=")[1])
            for valve in rest:
                edges[name].add(valve)
        for from_, to in edges.items():
            k = the_map[from_]
            v = {the_map[_] for _ in to}
            tunnels[k] = v
        return Cave(start, valves, rates, tunnels)

    def get_distances(self) -> list[list[int]]:
        size = len(self.valves)
        relevant_valves = [
            i for i in range(size) if self.rates[i] > 0 or i == self.start
        ]
        distances = [[0] * size for i in range(size)]
        for i in relevant_valves:
            result = Dijkstra.all(
                i,
                lambda _: False,
                lambda v: (_ for _ in self.tunnels[v]),
                lambda _, __: 1,
            )
            for j in relevant_valves:
                distances[i][j] = result.get_distance(j)
        return distances


class DFS:
    def __init__(self, max_time: int, *, sample: bool) -> None:
        self.max_time = max_time
        self.sample = sample
        self.used = 0
        self.max_flow = 0
        self.best_per_used = defaultdict[int, int](int)

    def dfs(
        self, cave: Cave, distances: list[list[int]], start: int, time: int
    ) -> None:
        for i in range(len(cave.valves)):
            idx = 1 << i
            if cave.rates[i] == 0 or self.used & idx != 0:
                continue
            new_time = time + 1 + distances[start][i]
            if new_time >= self.max_time:
                continue
            flow = cave.rates[i] * (self.max_time - new_time)
            if (
                not self.sample
                and self.max_flow + flow < self.best_per_used[self.used + idx]
            ):
                continue
            self.max_flow += flow
            self.used += idx
            self.dfs(cave, distances, i, new_time)
            self.max_flow -= flow
            self.used -= idx
        best = self.best_per_used[self.used]
        self.best_per_used[self.used] = max(best, self.max_flow)


Input = Cave
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Cave.from_input(list(input_data))

    def part_1(self, cave: Input) -> Output1:
        dfs = DFS(30, sample=False)
        dfs.dfs(cave, cave.get_distances(), cave.start, 0)
        return max(dfs.best_per_used.values())

    def solve_2(self, cave: Cave, *, sample: bool) -> int:
        dfs = DFS(26, sample=sample)
        dfs.dfs(cave, cave.get_distances(), cave.start, 0)
        return max(
            pressure1 + pressure2
            for used1, pressure1 in dfs.best_per_used.items()
            for used2, pressure2 in dfs.best_per_used.items()
            if used1 & used2 == 0
        )

    def part_2(self, cave: Input) -> Output2:
        return self.solve_2(cave, sample=False)

    @aoc_samples((("part_1", TEST, 1651),))
    def samples(self) -> None:
        cave = self.parse_input(TEST.splitlines())
        assert self.solve_2(cave, sample=True) == 1707


solution = Solution(2022, 16)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
